File: 146-lru-cache/lru-cache.py
class ListNode:
    def __init__(self, key, value):
        self.key = key
        self.val = value
        self.next = None
        self.prev = None

class LRUCache:

    def __init__(self, capacity: int):
        self.cache = {}
        self.capacity = capacity
        self.head = ListNode(-1, -1)
        self.tail = ListNode(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            # Move node to the tail (most recently used).
            self.remove(node)
            self.add(node)
            return node.val
        return -1        

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            old_node = self.cache[key]
            self.remove(old_node)
        
        node = ListNode(key, value)
        self.cache[key] = node
        self.add(node)

        if len(self.cache) > self.capacity:
            node_to_delete = self.head.next
            self.remove(node_to_delete)
            del self.cache[node_to_delete.key]

    def add(self, node: ListNode):
        previous_end = self.tail.prev
        previous_end.next = node
        node.prev = previous_end
        node.next = self.tail
        self.tail.prev = node

    def remove(self, node: ListNode):
        node.prev.next = node.next
        node.next.prev = node.prev

File: 146-lru-cache/lru-cache.java
class LRUCache {

    class Node {
        int key;
        int value;

        Node next;
        Node prev;

        public Node (int key, int value) {
            this.key = key;
            this.value = value;
            this.next = null;
            this.prev = null;
        }
    }

    private final int capacity;
    private final Map<Integer, Node> cache;
    private final Node head, tail;

    public LRUCache(int capacity) {

        this.cache = new HashMap<>();
        this.capacity = capacity;
        
        this.head = new Node(-1, -1);
        this.tail = new Node(-1, -1);

        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {

        if (cache.containsKey(key)) {
            Node currentNode = cache.get(key);
            remove(currentNode);
            insertInFront(currentNode);
            return currentNode.value;
        }
        return -1;
    }
    
    public void put(int key, int value) {

        if (cache.containsKey(key)) {
            Node currentNode = cache.get(key);
            remove(currentNode);
            cache.remove(key);
        }

        if (cache.size() >= capacity) {
            Node lastNode = tail.prev;
            remove(lastNode);
            cache.remove(lastNode.key);
        }

        Node newNode = new Node(key, value);
        cache.put(key, newNode);
        insertInFront(newNode);
    }

    public void remove(Node node) {

        node.prev.next = node.next;
        node.next.prev = node.prev;

    }

    public void insertInFront(Node node) {

        node.next = head.next;
        head.next.prev = node;
        node.prev = head;
        head.next = node;

    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */